crates/telio-proxy/src/proxy.rs

```diff
@@ -187,10 +208,61 @@ impl StateEgress {
 
         self.wg_addr = wg_port.map(|p| (Ipv4Addr::LOCALHOST, p).into());
 
-        self.sockets = sockets;
+        let old_sockets = self.sockets.clone();
+        self.sockets = sockets
+            .clone()
+            .into_iter()
+            .map(|(pk, socket)| {
+                (
+                    pk,
+                    (
+                        socket,
+                        old_sockets
+                            .get(&pk)
+                            .map(|(_, muted)| *muted)
+                            .unwrap_or(false),
+                    ),
+                )
+            })
+            .collect();
 
         Ok(())
     }
+
+    fn try_mute(&mut self, pk: &PublicKey, dur: Duration) -> Result<(), Error> {
+        if let Some((_, muted)) = self.sockets.get_mut(pk) {
+            telio_log_debug!("Muting peer: {:?} for {:?}", pk, dur);
+            self.mutes.insert(*pk, PinnedSleep::new(dur, *pk));
+            *muted = true;
+            return Ok(());
+        }
+
+        Err(Error::PeerNotFound)
+    }
+
+    fn try_unmute(&mut self, pk: &PublicKey) -> Result<(), Error> {
+        if let Some((_, muted)) = self.sockets.get_mut(pk) {
+            telio_log_debug!("Unmuting peer: {:?}", pk);
+            *muted = false;
+            self.mutes.remove(pk);
+            return Ok(());
+        }
+
+        Err(Error::PeerNotFound)
+    }
+
+    fn cleanup_mutes(&mut self) {
+        for (pk, mute) in self.mutes.iter() {
+            if mute.is_elapsed() {
+                // Cannot try_remove() here because of borrow checker
+                if let Some((_, muted)) = self.sockets.get_mut(pk) {
+                    telio_log_debug!("Cleaning mute from peer {:?}", pk);
+                    *muted = false;
+                }
+            }
+        }
+        self.mutes.retain(|_, mute| !mute.is_elapsed());
```

Comment:
`HashMap::retain` gives you mutable access to the values, so you can write it in a single loop

```suggestion
        for (pk, mute) in self.mutes.iter() {
            if mute.is_elapsed() {
                // Cannot try_remove() here because of borrow checker
                if let Some((_, muted)) = self.sockets.get_mut(pk) {
                    telio_log_debug!("Cleaning mute from peer {:?}", pk);
                    *muted = false;
                }
            }
        }
        self.mutes.retain(|_, mute| if mute.is_elapsed() {
                if let Some((_, muted)) = self.sockets.get_mut(pk) {
                    telio_log_debug!("Cleaning mute from peer {:?}", pk);
                    *muted = false;
                }
                true
            } else {
               false
        });
```

Reply:
look's good

Reply:
Here I don't think we can use hashmap::retain() in the way you suggested, self.mutes closure borrows it as immutable and then we borrow self.sockets as mutable. Borrow checker doesn't allows this

Reply:
Yeah, but this compiles fine:
```rust
        let socks = &mut self.sockets;

        self.mutes.retain(|pk, mute| {
            if mute.is_elapsed() {
                if let Some((_, muted)) = socks.get_mut(pk) {
                    telio_log_debug!("Cleaning mute from peer {:?}", pk);
                    *muted = false;
                }
                true
            } else {
                false
            }
        });
```

Reply:
I've just removed this function on the last push anw